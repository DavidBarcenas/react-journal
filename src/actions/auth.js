import { types } from "../types/types";

export const startLoginEmailPwd = ( email, pwd ) => {
  return ( dispatch ) => {
    setTimeout(() => {
      dispatch(login(123, 'Daveepro'))
    }, 3500);    
  }
}

export const login = ( uid, displayName ) => ({
  type: types.login,
  payload: {
    uid, 
    displayName
  }
})