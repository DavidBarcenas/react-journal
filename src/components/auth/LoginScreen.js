import React from 'react'
import { Link } from 'react-router-dom'
import { useDispatch } from 'react-redux'
import { useForm } from '../../hooks/useForm'
import { login } from '../../actions/auth'

export const LoginScreen = () => {

  const dispatch = useDispatch();

  const [ formValues, handleInputChange ] = useForm({
    email: 'davee@gmail.com',
    pwd: '123456'
  });

  const { email, pwd } = formValues;

  const handleLogin = (e) => {
    e.preventDefault();
    dispatch( login(12345, 'Davee') )
  }

  return (
    <>
      <h3 className="auth__title">Login</h3>

      <form onSubmit={ handleLogin }>
        <input type="email" 
          className="auth__input" 
          placeholder="Email" 
          name="email" 
          value={ email } 
          onChange={ handleInputChange }
        />
        <input type="password" 
          className="auth__input" 
          placeholder="Password" 
          name="pwd" 
          value={ pwd } 
          onChange={ handleInputChange } 
        />
        <button type="submit" className="btn btn-primary btn-block">Sign in</button>
      </form>

      <div>
        <span className="auth__social-text">Login with social networks</span>
        <button className="btn btn-block btn__google" >
          <div className="btn__google-img" >
            <img src="/images/google.svg" alt="google"/>
          </div>
          <span>Sign in with google</span>
        </button>

        <Link to="/auth/register" className="auth__social-text"> Create new account </Link>
      </div>
    </>
  )
}
