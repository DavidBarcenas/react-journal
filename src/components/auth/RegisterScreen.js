import React from 'react'
import { Link } from 'react-router-dom'
import { useForm } from '../../hooks/useForm'
import { useDispatch, useSelector } from 'react-redux';
import isEmail from 'validator/lib/isEmail';
import { setError, removeError } from '../../actions/ui';
import { startRegister } from '../../actions/auth';

export const RegisterScreen = () => {

  const dispatch = useDispatch();
  const { msgError } = useSelector(state => state.ui)

  /* DELETE INITIFORM FOR PRODUCTION ============== */
  const initialForm = {
    name: 'Max',
    email: 'max@mail.com',
    pwd: 'helloworld12',
    pwdConfirm: 'helloworld12',
  }

  const [ formValues, handleInputChange ] = useForm(initialForm);
  const { name, email, pwd, pwdConfirm } = formValues;

  const handleSubmit = (e) => {
    e.preventDefault();
    if( isFormValid() ) {
      dispatch( startRegister( email, pwd, name ) )
    }
  }

  const isFormValid = () => {
    if( name.trim().length === 0 ) {
      dispatch( setError('Name is required') )
      return false;
    } 
    else if( !isEmail( email ) ) {
      dispatch( setError('Email is not valid') )
      return false;
    } 
    else if( pwd !== pwdConfirm || pwd.length < 5 ) {
      dispatch( setError('Password should be at least six characters and match each other') )
      return false;
    }

    dispatch( removeError() )
    return true;
  }

  return (
    <>
      <h3 className="auth__title">Register</h3>

      { msgError && <div className="auth__alert-error"> { msgError} </div> }

      <form onSubmit={ handleSubmit }>
        <input type="text" className="auth__input" placeholder="Name" name="name" value={ name } onChange={ handleInputChange } />
        <input type="email" className="auth__input" placeholder="Email" name="email" value={ email } onChange={ handleInputChange } />
        <input type="password" className="auth__input" placeholder="Password" name="pwd" value={ pwd } onChange={ handleInputChange } />
        <input type="password" className="auth__input" placeholder="Password confirm" name="pwdConfirm" value={ pwdConfirm } onChange={ handleInputChange }/>
        <button type="submit" className="btn btn-primary btn-block">Register</button>
      </form>

      <Link to="/auth/login" className="auth__social-text"> Already registered?</Link>
    </>
  )
}
