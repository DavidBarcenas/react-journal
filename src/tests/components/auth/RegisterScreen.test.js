import '@testing-library/jest-dom';
import React from 'react'
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store'
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { RegisterScreen } from '../../../components/auth/RegisterScreen';
import { types } from '../../../types/types';

const middlewares = [thunk]
const mockStore = configureStore(middlewares)
const initState = {
  ui: {
    loading: false,
    msgError: null
  }
}
let store = mockStore( initState )

const wrapper = mount(
  <Provider store={ store }>
    <MemoryRouter>
      <RegisterScreen />
    </MemoryRouter>
  </Provider>
)

describe('<RegisterScreen /> testing', () => {
  
  test('should render correctly', () => {
    expect(wrapper).toMatchSnapshot()
  })
  
  test('should trigger the action', () => {
    const emailField = wrapper.find('input[name="email"]')
    emailField.simulate('change', {
      target: {
        value: '',
        name: 'email'
      }
    })
    wrapper.find('form').simulate('submit', {
      preventDefault(){}
    })
    const actions = store.getActions()
    expect(actions[0]).toEqual({ 
      type: types.uiSetError,
      payload: 'Email is not valid' 
    })
  })

  test('should show the alert box with the error', () => {
    const initState = {
      ui: {
        loading: false,
        msgError: 'Email is not valid'
      }
    }
    const store = mockStore( initState )
    
    const wrapper = mount(
      <Provider store={ store }>
        <MemoryRouter>
          <RegisterScreen />
        </MemoryRouter>
      </Provider>
    )

    expect(wrapper.find('.auth__alert-error').exists()).toBeTruthy()
    expect(wrapper.find('.auth__alert-error').text().trim()).toBe(initState.ui.msgError)
  })
  
})
