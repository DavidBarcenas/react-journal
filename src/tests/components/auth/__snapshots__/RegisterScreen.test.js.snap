// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RegisterScreen /> testing should render correctly 1`] = `
Array [
  <h3
    className="auth__title"
  >
    Register
  </h3>,
  <form
    onSubmit={[Function]}
  >
    <input
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value="Max"
    />
    <input
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="email"
      value="max@mail.com"
    />
    <input
      className="auth__input"
      name="pwd"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value="helloworld12"
    />
    <input
      className="auth__input"
      name="pwdConfirm"
      onChange={[Function]}
      placeholder="Password confirm"
      type="password"
      value="helloworld12"
    />
    <button
      className="btn btn-primary btn-block"
      type="submit"
    >
      Register
    </button>
  </form>,
  <a
    className="auth__social-text"
    href="/auth/login"
    onClick={[Function]}
  >
     Already registered?
  </a>,
]
`;
